File: Algorithm/Python/84.largest-rectangle-in-histogram.py
#
# @lc app=leetcode id=84 lang=python
#
# [84] Largest Rectangle in Histogram
#

# @lc code=start

class Solution(object):
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        result, stack = 0, []
        for i in range(len(heights)):
            while stack and heights[i] < heights[stack[-1]]:
                idx = stack.pop()
                if stack:
                    result = max(result, heights[idx] * (i - stack[-1] - 1))
                else:
                    result = max(result, heights[idx] * i)
            stack.append(i)
        
        # deal with increasing elements still in stack
        # including equal elements
        max_idx = stack[-1]
        while stack:
            idx = stack.pop()
            if stack:
                result = max(result, heights[idx] * (max_idx-stack[-1]))
            else:
                result = max(result, heights[idx] * (max_idx+1))
        return result
        
# @lc code=end

File: Algorithm/C++/84.largest-rectangle-in-histogram.cpp
/*
 * @lc app=leetcode id=84 lang=cpp
 *
 * [84] Largest Rectangle in Histogram
 */

// @lc code=start
#include <vector>
#include <stack>

class Solution {
public:
    int largestRectangleArea(std::vector<int>& heights) {
        // construct dummy heights
        std::vector<int> dummy_heights(heights.size()+2, 0);
        for (int i = 0; i < heights.size(); i++) {
            dummy_heights[i+1] = heights[i];
        }

        int curr_height, result = 0;
        std::stack<int> s;
        for (int i = 0; i < dummy_heights.size(); i++) {
            while(s.size() > 0 && dummy_heights[s.top()] > dummy_heights[i]) {
                curr_height = dummy_heights[s.top()]; 
                s.pop();
                result = std::max(result, curr_height * (i - s.top() - 1));
            }
            s.push(i);
        }
        return result;
    }
};
// @lc code=end

